Concurrent Session Control

fspmarshall opened the pull request:
Support for limiting concurrent sessions (#2938), implemented via new `semaphore` API.

*EDIT*: See updated PR description [here](#issuecomment-680273617).

<details>

<summary>Old PR Description</summary>

### TODOs

- Primary:
  - [x] Integration tests to verify concurrent session limits are correctly applied.
  - [x] Integration tests to verify termination of SSH connections when lease cannot complete before timeout.
  - [x] Unit testing of the `SemaphoreLock` abstraction.
  - [x] Audit event for users who attempt to exceed `max_concurrent_sessions`.
  - [x] Make `sessctl` lease expiry a cluster-level config option.

- Secondary:
  - [x] Unique error variant to trigger early `SemaphoreLock` release on failed renewal (useful for backwards compatibility in the event we end up using the semaphore system as the basis for Session Termination in the future).
  - [ ] Track kube proxy connections in addition to ssh connections (likely requires updating `github.com/gravitational/oxy/forward` to support cancellation).
  - [ ] Improve UX around SSH connections that are closed due to loss of a previously held lock (potentially difficult).

### Overview

The `role` resource has been updated to support a new `max_concurrent_sessions` field:

```yaml
kind: role
metadata:
  name: some-role
  # ...
spec:
  options:
    max_concurrent_sessions: 3
  # ...
version: v3
```

When a node encounters a new inbound SSH connection from a user who has a `max_concurrent_sessions` limit defined in one or more if their roles, it calculates the *minimum* of the defined limits as the maximum for that user.  The node will then attempt to  acquire a lease for the corresponding semaphore (`sessctl/<username>`) by invoking the auth server's semaphore API.  If the number of existent leases is already greater than or equal to the derived `max_concurrent_sessions` value, the ssh connection is rejected.  

When using `tsh`, the rejection looks like this:

```
$ tsh ssh alice@example.com
error: ssh: rejected: administratively prohibited (cannot acquire semaphore sessctl/alice (max leases reached))
```

When using `openssh`, the rejection looks like this:

```bash
$ ssh alice@example.com
channel 0: open failed: administratively prohibited: cannot acquire semaphore sessctl/alice (max leases reached)
Connection to example.com closed.
```

If a node *does* successfully acquire a semaphore lease, then the ssh connection continues normally, and the lease is periodically renewed in the background.

Semaphore leases have an expiry, and the node must successfully renew the lease before this expiry.  Nodes begin attempting to renew a lease when it is halfway to its expiry point.  If a node fails to successfully renew before the expiry time, the node will terminate the ssh connection.

~Existant leases can be viewed like so:~ (see [this](https://github.com/gravitational/teleport/pull/4138#issuecomment-672195474) comment for updated syntax)

```
$ tctl sem ls
Kind    Name  LeaseID                              Holder                               Expires             
------- ----- ------------------------------------ ------------------------------------ ------------------- 
sessctl alice a0585fc3-9770-44b7-b836-4d9c01751ee5 dc43a481-3063-4476-a9c7-505d1fffc077 28 Jul 20 00:35 UTC 
sessctl alice 0f3e3359-af4f-4a6b-9a36-ebb906e062c6 dc43a481-3063-4476-a9c7-505d1fffc077 28 Jul 20 00:35 UTC
```

Or, more verbosely:

```
$ tctl sem ls --format=json
[
  {
    "kind": "semaphore",
    "sub_kind": "sessctl",
    "version": "v3",
    "metadata": {
      "name": "alice",
      "expires": "2020-07-28T00:35:26.828932214Z"
    },
    "spec": {
      "leases": [
        {
          "lease_id": "a0585fc3-9770-44b7-b836-4d9c01751ee5",
          "expires": "2020-07-28T00:35:26.828932214Z",
          "holder": "dc43a481-3063-4476-a9c7-505d1fffc077"
        },
        {
          "lease_id": "0f3e3359-af4f-4a6b-9a36-ebb906e062c6",
          "expires": "2020-07-28T00:35:22.230492119Z",
          "holder": "dc43a481-3063-4476-a9c7-505d1fffc077"
        }
      ]
    }
  }
]
```

The semaphore API allows `lease_id` and `holder` to be arbitrary strings (though `lease_id` must be unique).  When a node acquires a semaphore lease for the purposes of limiting concurrent SSH connections, `lease_id` will be a random UUID, and `holder` will the UUID of the node in question.

~A pair of additional *hidden* `tctl` commands, `tctl sem cancel` and `tctl sem rm`, allow removal of individual leases and entire semaphores respectively.  These commands effectively constitute a form of asynchronous session termination, but do not prevent new sessions from being created.~ These two functions have been merged into a single `tctl sessctl rm` subcommand.  It is still hidden to avoid confusion since removing individual sessions may be needed in some debugging scenarios, but is not a substitute for proper session termination.

### Gotchas/Clarifications 

- Since Concurrent Session Control is applied at the authenticated SSH *connection* level, multiple individual "sessions" may appear in the audit log for each individual lease (in practice, this should only occur when connection reuse systems such as `ControlMaster` are in play).
- Semaphores do not store a global lease maximum as part of their state.  Instead, a lease acquisition attempt includes a parameter specifying the limit to be observed by the auth server that specific request.  This mechanism is designed to avoid problems with global configuration sync and unexpected behavior due to distributed ordering issues.  The result of this is that if state within a teleport cluster is out of sync, each node will individually refuse to create connections if the number of concurrent connections would exceed the `max_concurrent_sessions` value as seen by that node.
- Semaphore leases are renewed *only* by ID, and a single lease can theoretically be concurrently renewed by multiple separate entities (so long as the second entity is willing to overlook the `AlreadyExists` error).  This is an intentional design choice since some "session" concepts that we may want to track in the future won't exist on a unique machine (kubernetes API, AAP, etc...).

</details>

klizhentas committed:
Initial work on semaphores

benarent commented:
Thanks for the awesome PR Description. 

My initial gut reaction, is that we're leaking the implementation detail of semaphores to customers. It feels like a new concept, and I'm proposing for tctl we stick to sessions vs sems? 

```diff
# lists active sessions
$ tctl sem ls
+ $ tctl sessions ls

# Cancel vs rm? Aren't these the same?
$ tctl sem cancel
+ $ tctl session deny

# Removes / kills session
$ tctl sem rm 
+ $ tctl session rm
```

Also for `$ tctl sem ls` Can you find which node that users is accessing?

fspmarshall commented:
> My initial gut reaction, is that we're leaking the implementation detail of semaphores to customers. It feels like a new concept, and I'm proposing for tctl we stick to sessions vs sems?

You are absolutely right that we are leaking the semaphore concept... I'm hesitant to make a `tctl` command called `sessions` however, because the audit log has already given users an intuitive understanding of what a `tctl sessions ls` would return... and it doesn't quite map to the result of the current `tctl sem ls`.  The Session Control feature is limiting the number of concurrent authenticated connections, whereas the UI and audit log treat individual ssh exec/shell operations as sessions.

Even if we wanted to, I don't see how we could make the session construct used by the audit log match up with the [nist control](https://nvd.nist.gov/800-53/Rev4/control/AC-10) because the audit log's session construct doesn't have a 1-to-1 mapping to an authenticated user (due to the "session joining" feature).

We *could* create a `tctl sessctl ls` subcommand to provide a more tailored experience for specifically managing the Session Control feature, rather than using the general semaphore concept.  Perhaps something like this:

```
$ tctl sessctl ls
User  ID                                   Node                                 Expires             
----- ------------------------------------ ------------------------------------ ------------------- 
alice a0585fc3-9770-44b7-b836-4d9c01751ee5 dc43a481-3063-4476-a9c7-505d1fffc077 28 Jul 20 00:35 UTC 
alice 0f3e3359-af4f-4a6b-9a36-ebb906e062c6 dc43a481-3063-4476-a9c7-505d1fffc077 28 Jul 20 00:35 UTC
```

Do you think that is a better UX?  Definitely a little more self-explanatory.

---

> `# Cancel vs rm? Aren't these the same?`

Each semaphore contains some number of leases; if `alice` is under Session Control and has two currently active connections, then a semaphore named `sessctl/alice` will exist, and it will contain two leases (one for each connection).  The `cancel` operation removes a specific lease, whereas the `rm` operation removes the entire semaphore (which is equivalent to removing all leases).

Effectively, `cancel` will terminate a specific connection, whereas `rm` will terminate all sessions related to the user.  As it exists now, the `rm` operation can also be used to kill *all* connections in the cluster which are under Concurrent Session Control, but it won't let you do that unless you supply `-f/--force`.

If we went with the `tctl sessctl` subcommand strategy discussed above, we could probably unify these two concepts into a single subcommand:

```
$ tctl sessctl rm [--force] [--user=<username>] [<id>]...
```

---

> Also for `$ tctl sem ls` Can you find which node that users is accessing?

The `Holder` column contains the `UUID` of the node which holds the lease.  Using that value, one can run `tctl get nodes/<uuid>` to find out all the details of the node that the user is accessing.

benarent commented:
I edited contents of @fspmarshall based on our call yesterday, just some tweaks to the output of `tctl sessctl ls`.

fspmarshall committed:
fix racy logging & session channel rejection

fspmarshall commented:
Updated `tctl` subcommand:

Regular:
```
$ tctl sessctl ls
User  LeaseID                              Host 
----- ------------------------------------ ----------- 
alice 8c7ffdb8-6cf6-4688-943e-0694ac456302 example.com 
alice ac2fafef-2519-4e74-a6e4-6f7b44870fc3 example.com 
```
Verbose:
```
$ tctl sessctl ls -v
User  LeaseID                              NodeID                               Expires    
----- ------------------------------------ ------------------------------------ -------------------  
alice 8c7ffdb8-6cf6-4688-943e-0694ac456302 dc43a481-3063-4476-a9c7-505d1fffc077 11 Aug 20 18:46 UTC 
alice ac2fafef-2519-4e74-a6e4-6f7b44870fc3 dc43a481-3063-4476-a9c7-505d1fffc077 11 Aug 20 18:46 UTC 
```
Json:
```
$ tctl sessctl ls --format=json
[
  {
    "kind": "semaphore",
    "sub_kind": "sessctl",
    "version": "v3",
    "metadata": {
      "name": "alice",
      "expires": "2020-08-11T18:46:48.479388046Z"
    },
    "spec": {
      "leases": [
        {
          "lease_id": "8c7ffdb8-6cf6-4688-943e-0694ac456302",
          "expires": "2020-08-11T18:46:43.710496567Z",
          "holder": "dc43a481-3063-4476-a9c7-505d1fffc077"
        },
        {
          "lease_id": "ac2fafef-2519-4e74-a6e4-6f7b44870fc3",
          "expires": "2020-08-11T18:46:48.479388046Z",
          "holder": "dc43a481-3063-4476-a9c7-505d1fffc077"
        }
      ]
    }
  }
]
```

webvictim commented on lib/services/types.proto:
```diff
@@ -975,3 +985,86 @@ message TunnelConnectionSpecV2 {
+    ];
+}
+
+// SemaphoreLeaseRef identifies an existant lease.
```
nit: existent

webvictim commented on lib/srv/regular/sshserver.go:
```diff
@@ -849,6 +850,67 @@ func (s *Server) HandleRequest(r *ssh.Request) {
+	// lock, but if we lose the lock due to some error (e.g. poor connectivity
+	// to auth server) then cancellation propagates in the other direction.
+	go func() {
+		// TODO: If lock was lost due to error, find a way to propagate
+		// an error message to user.
```
TODO left in here - might be ok, but wanted to point it out.

fspmarshall replied:
Oops, changed to `TODO(fspmarshall)`.  We don't seem to have a good model for propagating connection-level errors right now, but I would like to develop one eventually, and I want to remember that this is one of the places that would benefit.

webvictim commented on tool/tctl/common/semaphore_command.go:
```diff
@@ -0,0 +1,168 @@
+func (c *SemaphoreCommand) Delete(client auth.ClientI) error {
+	if c.leaseID != "" {
+		if c.username == "" {
+			return trace.BadParameter("cannod resolve lease %q without username", c.leaseID)
```
nit: cannot

webvictim commented on tool/tctl/common/semaphore_command.go:
```diff
@@ -0,0 +1,168 @@
+	c.scList.Flag("user", "Filter output by username").StringVar(&c.username)
+
+	c.scDelete = sc.Command("rm", "Delete active session-control leases").Alias("delete").Hidden()
+	c.scDelete.Flag("force", "Permits unconstrained deletions").Short('f').BoolVar(&c.force)
```
I wonder if `--force` should be a hidden flag as it permits the user to do dangerous things? We (as support) could always tell them to do it when there's no alternative but maybe we want to prevent people from discovering it.

webvictim commented on lib/auth/proto/auth.proto:
```diff
@@ -248,4 +253,14 @@ service AuthService {
     rpc UpdateUser(services.UserV2) returns (google.protobuf.Empty);
     // DeleteUser deletes an exisitng user in a backend by username.
```
nit: existing

webvictim commented:
As a relative newbie to Go, I like this implementation and think it seems pretty clean. I'll leave the more hardcore reviewing to backend engineers.

fspmarshall commented:
retest this please

awly commented on lib/services/suite/suite.go:
```diff
@@ -1011,6 +1013,124 @@ func (s *ServicesTestSuite) ClusterConfig(c *check.C, opts ...SuiteOption) {
+// the same property holds for releasing semaphores; release operations are
+// best-effort and allowed to fail.
+func (s *ServicesTestSuite) SemaphoreContention(c *check.C) {
+	const locks int64 = 20
```
Since this is a stress test of sorts, can you bump it to ~1000?

fspmarshall replied:
Bumped to 5 separate iterations of 50 each.  The auth server starts returning "too much contention" errors at around 100, but that is by design... we don't want bad code to cause the auth server sit for long periods of time on hundreds of requests.  Contention only happens on a per-semaphore basis, so a user would need to start >100 separate SSH connections within a few seconds to hit this with `sessctl`.

awly commented on lib/services/suite/suite.go:
```diff
@@ -1011,6 +1013,124 @@ func (s *ServicesTestSuite) ClusterConfig(c *check.C, opts ...SuiteOption) {
+// acquisitions result in the correct number of successful acquisitions.
+func (s *ServicesTestSuite) SemaphoreConcurrency(c *check.C) {
+	const maxLeases int64 = 2
+	const attempts int64 = 20
```
same here, 10x the numbers to increase likelihood of picking up races

awly commented on lib/services/suite/suite.go:
```diff
@@ -1011,6 +1013,124 @@ func (s *ServicesTestSuite) ClusterConfig(c *check.C, opts ...SuiteOption) {
+	lock, err = services.AcquireSemaphoreLock(context.TODO(), cfg)
+	c.Assert(err, check.IsNil)
+
+	select {
+	case <-lock.Done():
+		c.Fatalf("Unexpected lock failure: %v", lock.Error())
+	case <-time.After(time.Millisecond * 1500):
+		c.Fatalf("timeout waiting for successful lock renewal")
+	case <-lock.Renewed():
+	}
```
wrap this in a loop and make sure renewal works >1 time
you may need to lower `cfg.Expiry` to ~100ms

awly commented on lib/srv/authhandlers.go:
```diff
@@ -133,33 +133,29 @@ func (h *AuthHandlers) CheckPortForward(addr string, ctx *ServerContext) error {
 func (h *AuthHandlers) UserKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
 	fingerprint := fmt.Sprintf("%v %v", key.Type(), sshutils.Fingerprint(key))
 
-	// as soon as key auth starts, we know something about the connection, so
```
are changes in this file relevant to the PR?
if not, please revert

fspmarshall replied:
required for tests to pass.  fixes a data race that was being triggered by some of the concurrent session tests.

awly commented on lib/utils/retry.go:
```diff
@@ -60,6 +60,8 @@ type Retry interface {
 	// could fire right away if Duration is 0
 	After() <-chan time.Time
+	// MaxPeriod returns maximum period between retries
+	MaxPeriod() time.Duration
```
This seems to be unused

awly commented on tool/tctl/common/semaphore_command.go:
```diff
@@ -0,0 +1,168 @@
+	force   bool
+	verbose bool
+
+	scList   *kingpin.CmdClause
+	scDelete *kingpin.CmdClause
```
nit: no need to prefix these with `sc`
how about `listCmd, deleteCmd`?

awly commented on tool/tctl/common/semaphore_command.go:
```diff
@@ -0,0 +1,168 @@
+		}
+		_, err := table.AsBuffer().WriteTo(os.Stdout)
+		return trace.Wrap(err)
+	default:
```
```
case format == teleport.JSON:
  ...
default:
  return trace.BadArgument("unknown format %q", format)

fspmarshall commented on lib/auth/api.go:
```diff
@@ -156,6 +160,7 @@ func NewWrapper(writer AccessPoint, cache ReadAccessPoint) AccessPoint {
 // and read operations are going though cache
 type Wrapper struct {
 	ReadAccessPoint
+	services.Semaphores
 	Write AccessPoint
 }
```
So splitting the read and write halves here doesn't work like it does for the the other members of `AccessPoint` because we don't cache semaphores.  This is deliberate because semaphores are frequently modified, and all normal (non-`tctl`) semaphore operations are modifying operations, meaning that caching the semaphores would be a net loss in terms of performance.

The only reason `Wrapper` doesn't embed `AccessPoint` is because `AccessPoint` embeds `ReadAccessPoint`.  `Wrapper` is intercepting the read-only subset of methods on `AccessPoint` to divert them to the cache, and manually re-exporting the individual uncached methods.

Way I see it there are a few options here:
1. Don't embed `Semaphores`, and instead manually re-export all its methods (possibly renaming the `Write` field to `NoCache` in order to help clarify the real distinction here, since `Write.GetSemaphores(...)` is confusing).
2. Create a third `NoCacheAccessPoint` interface that is embedded in both `Wrapper` and `AccessPoint` just like `ReadAccessPoint` is, continuing to only manually export the methods which come from partially cached services (this is probably the most correct way to do things, but feels a bit silly since `Semaphores` would currently be the only member of `NoCacheAccessPoint`).
3. Keep it as is and add comment explaining that uncached services don't need to be subselected the way partially cached services do.  This is functionally the same as option 2, but possibly less convenient if we end up adding more uncached services down the road. 

Thoughts?

awly replied:
no. 1 sounds great, it makes the intention here clearer.
maybe coupled with renaming `ReadAccesPoint` to `CacheableAccessPoint` or similar.

russjones commented on lib/services/semaphore.go:
```diff
@@ -0,0 +1,604 @@
+
+// SemaphoreKindSessionControl is the semaphore kind used by
+// the Concurrent Session Control feature.
+const SemaphoreKindSessionControl = "sessctl"
```
Do we have any other kind of semaphore?

russjones replied:
Also, probably need a better name for this @benarent.

benarent replied:
After today we are going to remove this for now, and review when we come back to locking / terminating sessions.

russjones commented on lib/services/semaphore.go:
```diff
@@ -0,0 +1,604 @@
+
+// AcquireSemaphoreLock attempts to acquire and hold a semaphore lease.  If successfully acquired,
+// background keepalive processes are started and an associated lock handle is returned.
+func AcquireSemaphoreLock(ctx context.Context, cfg SemaphoreLockConfig) (*SemaphoreLock, error) {
```
I think this function should live in `lib/client/semaphore.go` and be callable from the Auth Server client. We don't have a whole lot of application logic that sits in `lib/services`, it's mainly data structures, so it feels out of place. @fspmarshall what do you think?

fspmarshall replied:
Good point.  Unfortunately, it can't go in `client` due to circular import issues (`client` depends on `auth` depends on `services/suite`  depends on `AcquireSemaphoreLock`).  I kept it in `services` because keeping semaphore stuff together seemed like a reasonable default, but I'm open to alternatives.

russjones replied:
Can you use an interface to break the circular dependencies?

fspmarshall replied:
Maybe... I could make a `SemaphoreLock` interface and an `AcquireSemaphoreLockFunc` type and move the concrete impls somewhere else.  `SemaphoreLockConfig` would still need to be defined in services, and wherever we moved the concrete implementation, it would still need to be importable  from `services/suite`.  We could put it in `services/local` if we made `services/suite` accept a `SemaphoreLockFunc` parameter, but thats a lot of indirection for something that doesn't actually need to have multiple concrete implementations.

I think `SemaphoreLock` falls into the same category as things like `services.ProxyWatcher` and `services.Fanout`.  Simple self-contained helpers which are technically application logic, but basically just wrap the interfaces defined in `services` and make them a bit easier to use in certain ways.  Maybe we should create something like a `services/helpers` for things like this to live in?

russjones replied:
Okay, let's pass on the interface then.

fspmarshall commented on lib/services/semaphore.go:
```diff
@@ -0,0 +1,604 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctx, cancel := context.WithCancel(ctx)
```
Being able to limit the lifespan of the lock's keepalive operations to a specific context is really handy, but I do see the issue with confusing the acquire context and the keepalive context.  Refactored lock semantics to look like this:

```golang
lock, _ := AcquireSemaphoreLock(ctx, ...)
go lock.KeepAlive(ctx)
```

Now if a caller desires they can use separate contexts for the initial acquire and the long-running keepalive.

fspmarshall commented on lib/services/local/presence.go:
```diff
@@ -657,6 +658,300 @@ func (s *PresenceService) DeleteAllRemoteClusters() error {
+	// key is not modified, so allocate it once
+	key := backend.Key(semaphoresPrefix, req.SemaphoreKind, req.SemaphoreName)
+
+	for i := int64(0); i < acquireAttempts; i++ {
```
I'm using `int64` here because `int64` is the base type of `time.Duration`.  Technically, it isn't necessary in this case, but I don't generally cast between numeric types unless I have some compelling reason to do so.

fspmarshall commented on lib/services/local/presence.go:
```diff
@@ -657,6 +658,300 @@ func (s *PresenceService) DeleteAllRemoteClusters() error {
+
+// KeepAliveSemaphoreLease updates semaphore lease, if the lease expiry is updated,
+// semaphore is renewed
+func (s *PresenceService) KeepAliveSemaphoreLease(ctx context.Context, lease services.SemaphoreLease) error {
```
Yes.  Unlike `AcquireSemaphore`, these operations let the caller decide how to handle retries.  There are two main reasons for this difference (in no particular order):

- `Acquire` blocks user-facing operations, whereas `KeepAlive`/`Cancel` are handled asynchronously in the background (i.e. `Acquire` benefits much more from fast retries).
- `Acquire` abstracts over multiple different possible contention events, whereas `KeepAlive`/`Cancel` represent contention in the "normal" way, by propagating a `CompareFailed` from the backend (i.e. bigger payoff in terms of API simplicity when encapsulating `Acquire`'s retry model).

I'll update the method comments to better reflect this.

fspmarshall commented on lib/services/types.proto:
```diff
@@ -975,3 +985,86 @@ message TunnelConnectionSpecV2 {
+}
+
+// SemaphoreLeaseRef identifies an existent lease.
+message SemaphoreLeaseRef {
```
`SemaphoreLease` specifies the name and kind of the semaphore, which `SemaphoreLeaseRef` does not need to store since `SemaphoreLeaseRef` is stored within its semaphore.  Meanwhile, `SemaphoreLeaseRef` stores `Holder`, which `SemaphoreLease` does not need to store since `Holder` is basically just debug info.

Generally, the two types have different requirements, so it seemed better to define them separately from the beginning.

awly replied:
I guess it's a tradeoff between minimizing storage (purpose-built minimal objects) and schema simplicity (reusable objects where some fields are occasionally unused).
I like the latter because both `SemaphoreLease` and `SemaphoreLeaseRef` represent the same logical "thing".
But if you and @russjones both prefer the current approach, I'm OK with it.

fspmarshall commented on lib/services/local/presence.go:
```diff
@@ -657,6 +658,300 @@ func (s *PresenceService) DeleteAllRemoteClusters() error {
+}
+
+// DeleteSemaphores deletes all semaphores matching the supplied filter
+func (s *PresenceService) DeleteSemaphores(ctx context.Context, filter services.SemaphoreFilter) error {
```
Actually, we just never delete semaphores.  They expire automatically.  The `Delete` method is only used in tests and debugging with `tctl`.  That being said, I'll tweak this method to require that semaphore kind is specified, which will prevent people from accidentally passing zeroed filters.  Think that is sufficient?

awly replied:
I'm in the YAGNI camp on this one - let's not implement multiple delete until there's a use case.

fspmarshall commented on lib/services/semaphore.go:
```diff
@@ -0,0 +1,604 @@
+	if c.Expiry().IsZero() {
+		return trace.BadParameter("set semaphore expiry time")
+	}
+	if c.SubKind == "" {
+		return trace.BadParameter("supply semaphore SubKind parameter")
+	}
```
`Name` is checked during `Metadata.CheckAndSetDefaults` call above.

fspmarshall commented on lib/services/types.proto:
```diff
@@ -975,3 +985,86 @@ message TunnelConnectionSpecV2 {
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "expires"
+    ];
```
Removed `LeaseID`.  Expiration is always decided by the node ahead of time tho.

fspmarshall committed:
concurrent session control

Adds support for Concurrent Session Control and a new
semaphore API.  Roles now support two new configuration
options, `max_ssh_connections` and `max_ssh_sessions`
which correspond to the total number of authenticated
ssh connections per cluster, and the number of ssh sessions
within a connection respectively.  Attempting to exceed
these limits generate variants of the `session.rejected`
audit event and cause the connection/session to be
rejected.

fspmarshall commented:
## Updated PR Description

### RBAC

A role's `options` block now supports two new optional configuration values:

```yaml
kind: role
metadata:
  name: some-role
  # ...
spec:
  options:
    max_ssh_connections: 2
    max_ssh_sessions: 2
  # ...
version: v3
```

---

#### `max_ssh_connections`

This value limits the total number of concurrent SSH connections that a user may establish to nodes within a cluster.  No distinction is made between connections to the same node or different node, and only connections to *teleport* nodes are covered (connections to `openssh` nodes are not counted).

In order to enforce this control, nodes acquire semaphore leases from the auth server corresponding to the teleport username that is attempting to connect.  This means that users who have limits placed on concurrent connections cannot connect to nodes while the auth server is offline.

In `tsh`, attempting to exceed `max_ssh_connections` will result in an error that looks something like this:

```
$ tsh ssh node.example.com
error: ssh: rejected: administratively prohibited (too many concurrent ssh connections for user "alice" (max=2))
```

The same error in `openssh` will look like this:

```
ssh -J proxy.example.com node.example.com
channel 0: open failed: administratively prohibited: too many concurrent ssh connections for user "alice" (max=2)
```

#### `max_ssh_sessions`

This value limits the total number of `session` channels which can be established across a single SSH connection (typically used for interactive terminals or remote `exec` operations).  It is essentially equivalent to the `MaxSessions` configuration value accepted by `sshd`.  Under normal usage via `tsh` a user will never hit this limit as separate invocations of `tsh` produce separate connections.  This limit will affect users or software that leverage connection-reuse systems such as the `ControlMaster` feature of `openssh`.

In `openssh`, attempting to exceed `max_ssh_sessions` will result in an error message something like this:

```
channel 5: open failed: administratively prohibited: too many session channels for user "alice" (max=2)
```

---

### Cluster Configuration

A new `session_control_timeout` configuration value has been added to the `auth_service` configuration block of the `teleport` config file:

```yaml
auth_service:
  session_control_timeout: 2m # default
# ...
```

This parameter controls the expiration time of the semaphore leases used to enforce `max_concurrent_connections`.  More generally, this value represents the upper limit of how long a node can go without being able to successfully contact the auth server before it terminates a controlled connection.  Because semaphores are synchronized cluster-wide state, very short timeouts may be costly performance-wise.  The default value is two minutes, but clusters with extremely high numbers of concurrent controlled connections may benefit from longer timeouts.

---

### The Semaphore Resource

In order to facilitate basic debugging/introspection, support for the semaphore resource has been added to the `tctl get` and `tctl rm` commands:

```
$ tctl get semaphores
kind: semaphore
metadata:
  expires: "2020-08-25T21:05:31.846078263Z"
  name: alice
spec:
  leases:
  - expires: "2020-08-25T21:05:31.846078263Z"
    holder: dc43a481-3063-4476-a9c7-505d1fffc077
    lease_id: bf36059c-7986-4f44-937e-a43cbda2b1f5
  - expires: "2020-08-25T21:05:30.726874902Z"
    holder: dc43a481-3063-4476-a9c7-505d1fffc077
    lease_id: b488bcf3-d6fa-4e31-a363-08ea1488de5d
sub_kind: ssh-conn
version: v3

$ tctl rm semaphores/ssh-conn/alice
semaphore 'ssh-conn/alice' has been deleted
```

### Audit Events

Attempts to exceed the `max_ssh_connections` and `max_ssh_sessions` restraints result in `ssh_connection` and `ssh_session` variants of the `sessctl.limit` audit event respectively:

```json
{
    "code": "T1006W",
    "event": "sessctl.limit",
    "max": 2,
    "proto": "ssh",
    "server_id": "dc43a481-3063-4476-a9c7-505d1fffc077",
    "sessctl_kind": "ssh_connection",
    "time": "2020-08-25T20:55:28.548Z",
    "uid": "adfad7cd-e884-47dd-8663-4d83de30f7cc",
    "user": "alice"
}
```
```json
{
    "code": "T1006W",
    "event": "sessctl.limit",
    "max": 2,
    "proto": "ssh",
    "server_id": "dc43a481-3063-4476-a9c7-505d1fffc077",
    "sessctl_kind": "ssh_session",
    "time": "2020-08-25T20:55:28.498Z",
    "uid": "9ad61a77-de3d-44c8-bafa-dc21e63a3150",
    "user": "alice"
}
```

klizhentas commented:
@fspmarshall does it have to be `ssh` specific? Can we have just `max_connections` and `max_sessions` that will limit total overall amount of connections and sessions across different protocol? Otherwise we would have to add a new option per protocol?

fspmarshall commented:
> @fspmarshall does it have to be `ssh` specific? Can we have just `max_connections` and `max_sessions` that will limit total overall amount of connections and sessions across different protocol? Otherwise we would have to add a new option per protocol?

@klizhentas I started off using the more general terminology, but was concerned that it might cause confusion since not all `connection`/`session` concepts have the hierarchical relationship that ssh connections/sessions have (i.e. a session is scoped to a specific connection).  For things like the kubernetes API and (I think) AAP, a single session can span multiple connections.  What do you think?  I suppose as long as we don't promise to *preserve* the current hierarchical relationship, it isn't a huge deal :/

klizhentas commented:
I would still remove ssh part, precisely because we don't promise to preserve the relationship, but we would actually limit the amount of connections and sessions across all ssh/k8s/app protocols - that's AC requirement anyways.

fspmarshall commented:
Updated [description](#issuecomment-680273617) to reflect removal of `ssh` prefixing.

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -304,6 +304,32 @@ func (s *SrvSuite) TestAgentForwardPermission(c *C) {
 
+// TestMaxSesssions makes sure that MaxSessions RBAC rules prevent
+// too many concurrent sessions.
+func (s *SrvSuite) TestMaxSessions(c *C) {
```
also check that the limit doesn't affect new connections

awly commented on lib/sshutils/ctx.go:
```diff
@@ -143,6 +147,30 @@ func (c *ConnectionContext) GetForwardAgent() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.sessions >= max {
+		return nil, false
```
to make this fool-proof
```suggestion
		return func() {}, false
```

awly commented on lib/sshutils/server.go:
```diff
@@ -490,6 +525,19 @@ func (f NewChanHandlerFunc) HandleNewChan(ctx context.Context, ccx *ConnectionCo
 	f(ctx, ccx, ch)
 }
 
+// NewConnHandler is called once per incoming connection.
+// Errors terminate the incoming connection.  The returned context
+// must be the same as, or a child of, the passed in context.
+type NewConnHandler interface {
+	HandleNewConn(ctx context.Context, ccx *ConnectionContext) (context.Context, error)
+}
+
+type NewConnHandlerFunc func(ctx context.Context, ccx *ConnectionContext) (context.Context, error)
+
+func (f NewConnHandlerFunc) HandleNewConn(ctx context.Context, ccx *ConnectionContext) (context.Context, error) {
+	return f(ctx, ccx)
+}
```
how about getting rid of the interface here and just using a callback func?

awly commented on tool/tctl/common/collection.go:
```diff
@@ -642,3 +642,47 @@ func (c *remoteClusterCollection) toMarshal() interface{} {
+	return trace.Wrap(err)
+}
+
+func (c *semaphoreCollection) toMarshal() interface{} {
+	if len(c.sems) == 1 {
+		return c.sems[0]
+	}
+	return c.sems
+}
```
if we format 1 and multiple semaphores differently, parsing structured output (json or yaml) will get difficult.
how about always formatting as an array?

fspmarshall replied:
In theory I agree, but this is how `tctl` already handles every other resource type, and I think consistency is more important.  If we want to change `tctl` to always output sequences, its probably better to do it in a separate PR.

awly reviewed (approved):
Note: I didn't re-review `lib/services/semaphore.go`, hit my mental capacity after ~2hrs of reviewing.

Regarding `max_connections` vs `max_sessions`: it looks like these separate controls are required by FIPS, right?
I suspect this will be confusing for some customers, vs a single _global_ `max_sessions`-like limit.
We should at least make the docs very clear about how these relate @benarent

fspmarshall commented on lib/services/types.proto:
```diff
@@ -975,3 +989,88 @@ message TunnelConnectionSpecV2 {
+
+// SemaphoreV3 implements Semaphore interface
+message SemaphoreV3 {
+    option (gogoproto.goproto_stringer) = false;
+    option (gogoproto.stringer) = false;
```
These options are set for all top-level `Resource` types.  We implement custom `String()` methods that display compact summaries of the important state of the resource.

russjones commented on lib/events/api.go:
```diff
@@ -326,6 +326,10 @@ const (
+
+	// SessionControlLimit fires when a user's attempt to create an authenticated
+	// session has been rejected due to exceeding `max_concurrent_sessions`.
+	SessionControlLimitEvent = "sessctl.limit"
```
Was this named to something else?

fspmarshall committed:
fix race in TestWeb

fspmarshall commented:
retest this please
